### PR TITLE
793 restructuring and reorganization of verification setups

### PR DESCRIPTION
**Pull request by jhabriel**
## Proposed changes

This PR does not introduce new functionality, but rather provides uniformity and solves several TODOs related to verification setups left from last year.  

### Main changes:

- The `verification_setup` directory was moved from `porepy.models` to `porepy.applications`.
- Material constants are now defined at the module level and not in the __init__() of the `SolutionStrategy`.
- Sanity checks were introduced for mandatory material constants, e.g., `compressibility = 0` for the fluid in the incompressible flow verification.
- `DataSavingMixin` is now employed to store results. The class `StoreResults` was purged from all verification setups.
- A small bug in `manu_flow_comp_frac.py` was fixed.
- Code duplication between the compressible and incompressible flow verifications was considerably reduced.
- Documentation was improved.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply_

- [ ] Minor change (e.g., dependency bumps, broken links, etc).
- [x] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code, etc).
- [ ] Other: 

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.	
- [x] Static typing is included in the update.
- [x] This PR does not duplicated existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).

### CODE REVIEW

File: src/porepy/applications/verification_setups/manu_flow_comp_frac.py [hunk 1]
```diff
@@ -846,114 +754,68 @@ def before_nonlinear_loop(self) -> None:
...
-    ModifiedGeometry,
-    ModifiedBoundaryConditions,
-    ModifiedBalanceEquation,
-    ModifiedSolutionStrategy,
+class ManufacturedCompressibleFlow2d(  # type: ignore[misc]
+    ManuCompBalanceEquation,
+    ManuCompBoundaryConditions,
+    ManuCompSolutionStrategy,
     pp.fluid_mass_balance.SinglePhaseFlow,
+    SetupUtilities,
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_comp_frac.py:779]: This looks dangerous. Usually, you want your mixin classes to have precedence over the base (SinglePhaseFlow).

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 2]
```diff
@@ -1,27 +1,99 @@
...
+
+    exact_rock_flux: np.ndarray = np.zeros(1)
+    """Exact flux in the rock."""
+
+    exact_rock_pressure: np.ndarray = np.zeros(1)
+    """Exact pressure in the rock."""
+
+
+class ManuIncompExactSolution:
     """Parent class for the exact solution."""
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:97]: In what sence "parent"?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 3]
```diff
@@ -338,85 +410,192 @@ def boundary_values(self, sd_rock: pp.Grid) -> np.ndarray:
...
+            true_array=out.exact_intf_flux,
+            approx_array=out.approx_intf_flux,
             is_scalar=True,
             is_cc=True,
         )
 
+        return out
+
+
+class SetupUtilities:
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:530]: Should this be moved? E.g. to verification_utils?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 4]
```diff
@@ -477,188 +657,192 @@ def set_md_grid(self) -> None:
...
-        sd_rock = self.mdg.subdomains()[0]
-        p_num = self.sol.approx_rock_pressure
-        p_ex = self.sol.exact_rock_pressure
-        pp.plot_grid(
-            sd_rock, p_ex, plot_2d=True, linewidth=0, title="Rock pressure (Exact)"
-        )
-        pp.plot_grid(
-            sd_rock, p_num, plot_2d=True, linewidth=0, title="Rock pressure (MPFA)"
-        )
+        self.subdomain_darcy_flux_variable: str = "darcy_flux"
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:767]: Why is this needed as a variable?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 5]
```diff
@@ -477,188 +657,192 @@ def set_md_grid(self) -> None:
...
+        self, solution: np.ndarray, errors: float, iteration_counter: int
+    ) -> None:
+        """Method to be called after if solution has converged."""
 
-        # TODO: This is ugly. Is there a better way to do this?
-        # Compute Darcy fluxes and store them in pp.STATE
-        darcy_flux = self.darcy_flux(self.mdg.subdomains())
-        vals = darcy_flux.evaluate(self.equation_system).val
-        nf_rock = self.mdg.subdomains()[0].num_faces
+        # Store subdomain Darcy fluxes in pp.STATE
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:790]: Why?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 6]
```diff
@@ -477,188 +657,192 @@ def set_md_grid(self) -> None:
...
         for sd, data in self.mdg.subdomains(return_data=True):
-            if sd.dim == 2:
-                data[pp.STATE]["darcy_flux"] = vals[:nf_rock]
-            else:
-                data[pp.STATE]["darcy_flux"] = vals[nf_rock:]
+            darcy_flux_ad = self.darcy_flux([sd])
+            vals = darcy_flux_ad.evaluate(self.equation_system).val
+            data[pp.STATE]["darcy_flux"] = vals
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:794]: And why is the name hard-coded here (not self.subdomain_darcy_flux_variable)?

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py [hunk 7]
```diff
@@ -362,159 +466,162 @@ def flow_source(self, sd: pp.Grid, time: float) -> np.ndarray:
...
+            true_array=out.exact_force,
+            approx_array=out.approx_force,
             is_scalar=False,
             is_cc=False,
         )
 
+        return out
+
+
+class SetupUtilities:
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_poromech_nofrac.py:582]: Again, move to verification_utils? Any possibility of refactoring for code reuse?

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py [hunk 8]
```diff
@@ -362,159 +466,162 @@ def flow_source(self, sd: pp.Grid, time: float) -> np.ndarray:
...
+        pp.plot_grid(sd, u_num[::2], plot_2d=True, linewidth=0, title="u_x (Numerical)")
+
+        # Vertical displacement
+        pp.plot_grid(sd, u_ex[1::2], plot_2d=True, linewidth=0, title="u_y (Exact)")
+        pp.plot_grid(
+            sd, u_num[1::2], plot_2d=True, linewidth=0, title="u_y (Numerical)"
+        )
+
+
+class UnitSquareTriangleGrid(pp.ModelGeometry):
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_poromech_nofrac.py:624]: The Geometry classes should be refactored to a applications/geometries. This will also reveal potential for reuse, I think.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py [hunk 9]
```diff
@@ -595,49 +717,71 @@ def body_force(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
...
+        """Results object that stores exact and approximated solutions and errors."""
 
-    def set_materials(self):
-        """Set material parameters.
+        self.flux_variable: str = "darcy_flux"
+        """Keyword to access the Darcy fluxes."""
 
-        Add exact solution object to the simulation model after materials have been set.
-        """
+        self.stress_variable: str = "poroelastic_force"
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_poromech_nofrac.py:770]: I don't like adding "variables" which are not primary and then storing them in STATE. This obscures what are really primary variables imo.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py [hunk 10]
```diff
@@ -652,94 +796,64 @@ def after_nonlinear_convergence(
...
     ModifiedEquationsPoromechanics,
     ModifiedSolutionStrategy,
     poromechanics.Poromechanics,
+    SetupUtilities,
+    VerificationUtils,
+    ModifiedDataSavinMixin,
 ):
     """
-    Mixer class for the verification setup.
+    Mixer class for the two-dimensional non-linear poromechanics verification setup.
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_poromech_nofrac.py:832]: Perhaps it suffices to point to the verification test file for usage example? Advantage is reduced maintenance (this example will be outdated at some point without us noticing).

File: src/porepy/applications/verification_setups/terzaghi_biot.py [hunk 11]
```diff
@@ -46,7 +46,7 @@
 import porepy.models.fluid_mass_balance as mass
 import porepy.models.poromechanics as poromechanics
 from porepy.applications.classic_models.biot import BiotPoromechanics
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/terzaghi_biot.py:48]: I suggest renaming from classic_models to derived_models.

File: tests/functional/test_manu_flow_comp_frac.py [hunk 12]
```diff
@@ -20,36 +24,51 @@
 the fracture) and fluxes (for the rock, the fracture, and on the interface). The
 errors are measured in a discrete relative L2-error norm (such as the ones defined
 in [2]). The desired errors were obtained by running the
-::class::ManufacturedCompFlow2d simulation setup with default parameters. We test the
-errors for three different times, namely: 0.2 [s], 0.6 [s], and 1.0 [s].
+:class:`ManufacturedCompressibleFlow2d` simulation setup with the parameters given in the
```
Comment (IvarStefansson) [inline tests/functional/test_manu_flow_comp_frac.py:27]: Line too long?

**Review by IvarStefansson — CHANGES_REQUESTED** [review]
Nice work!

I have some suggestions for refactoring. I noticed multiple modified DataSavingMixin overriding save_data_time_step. This could be solved by providing a VerificationDataSaving. Please do a search for Modified and replace with more specific class names. Otherwise, see specific comments.

**Comment by IvarStefansson** [comment]
Nice work!

I have some suggestions for refactoring. I noticed multiple modified DataSavingMixin overriding save_data_time_step. This could be solved by providing a VerificationDataSaving. Please do a search for Modified and replace with more specific class names. Otherwise, see specific comments.

File: src/porepy/applications/verification_setups/manu_flow_comp_frac.py [hunk 13]
```diff
@@ -158,39 +312,7 @@ def rock_pressure(self, sd_rock: pp.Grid, time: float) -> np.ndarray:
...
-
-        # Cell-centered densities
-        rho_cc = np.zeros(sd_rock.num_cells)
-        for (rho, idx) in zip(rho_fun, cell_idx):
-            rho_cc += rho(cc[0], cc[1], time) * idx
-
-        return rho_cc
-
-    def rock_darcy_flux(self, sd_rock: pp.Grid, time: float) -> np.ndarray:
+    def rock_flux(self, sd_rock: pp.Grid, time: number) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_comp_frac.py:315]: 1. This is Darcy flux in $m^3 s^{-1}$.
2. Sure, will rename.

[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/applications/verification_setups/manu_flow_comp_frac.py:315]: I'm not sure you want to rename, because this applies to a _lot_ of attributes/methods, I think

[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_comp_frac.py:315]: This is quite easy with PyCharm actually. It automatically changes all instances (including comments!).

[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_comp_frac.py]: Done.

File: src/porepy/applications/verification_setups/manu_flow_comp_frac.py (hunk 1 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_comp_frac.py:780]: True. Fixed.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 2 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:97]: Yeah. Bad wording here. Rephrased.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 4 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:767]: The idea here is to give keywords to secondary variables, e.g., subdomain Darcy fluxes. These are retrieved after convergence and stored in pp.STATE. Later, they are accessed by the data-saving classes.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 5 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:790]: Simply to be able to access them in the data-saving stage. Do you have another suggestion?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 6 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:794]: This should no be hardcoded, the keyword must have been used.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 8 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:624]: I agree. Shall we postpone this until we have a clear plan on how to implement this? Adaptation should be straightforward for the verification setups.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 9 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:770]: I don't particularly think that it will obscure what are the primary variables. After all, the user should be minimally familiar with MPFA/MPSA methods and know that the primary variables are pressure and displacement. Even if that's not the case, it should evident from the verification that these variables were added in a post-processing stage.

That being said, I can add a comment explaining that these are not primary variables. What do you think?

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 10 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:832]: I agree, although, from the user side, it is extremely convenient to have something that can be pasted into another file and run without having to inspect a test. For example, Mandel's verification setup will, by default, replicate the results of the PorePy paper (this of course is not feasible for a light-weighted test), which has an added value in itself.

My vote goes for keeping the Examples section, even though it adds a slight maintenance overhead.

File: src/porepy/applications/verification_setups/terzaghi_biot.py (hunk 11 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/terzaghi_biot.py]: Done.

File: tests/functional/test_manu_flow_comp_frac.py (hunk 12 above)
Comment (jhabriel) [reply tests/functional/test_manu_flow_comp_frac.py:27]: Yes. Fixed.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py [hunk 14]
```diff
@@ -595,49 +717,71 @@ def body_force(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
 
 
 class ModifiedEquationsPoromechanics(ModifiedMassBalance, ModifiedMomentumBalance):
-    ...
+    """Mixer class for modified poromoechanics equations."""
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:720]: Yes. Fixed.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 5 above)
Comment (IvarStefansson) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:790]: Could you compute them directly in data-saving stage?

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 8 above)
Comment (IvarStefansson) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:624]: I would vote for doing it straight away. But I guess I could be convinced to postpone.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 10 above)
Comment (IvarStefansson) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:841]: Let's allow the others to weigh in on this one. Opinions, @keileg, @OmarDuran?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 5 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:790]: Sure.

Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py]: Done for all verifications.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 9 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:751]: I went for your approach. Fixed.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 4 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:747]: Fixed.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 3 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py]: I renamed the classes and recycled utilities among verification setups. However, I don't think it is a good idea to use generic plotting methods.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 7 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py]: Did not move to verification_utils, but reused code when possible.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 10 above)
Comment (keileg) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:801]: I will need a short briefing before making an opinion.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 3 above)
Comment (IvarStefansson) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py]: I agree.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 15]
```diff
@@ -43,58 +46,160 @@
...
+    darcy_flux: Callable[[list[pp.Grid]], pp.ad.Operator]
+    """Method that returns the Darcy fluxes in the form of an Ad operator. Usually
+    provided by the mixin class :class:`porepy.models.constitutive_laws.DarcysLaw`.
+
+    """
+
+    exact_sol: ManuIncompExactSolution
+    """Exact solution object."""
+
+    interface_darcy_flux_variable: str
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:112]: Depracated?

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 5 above)
Comment (IvarStefansson) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py]: Does that mean that the secondary variable names should not be used? If yes, please purge them.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py [hunk 16]
```diff
@@ -716,52 +709,49 @@ def body_force(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
         return external_sources
 
 
-class ModifiedEquationsPoromechanics(ModifiedMassBalance, ModifiedMomentumBalance):
+class ManuPoroMechEquationsPoromechanics(
```
Comment (IvarStefansson) [inline src/porepy/applications/verification_setups/manu_poromech_nofrac.py:712]: The name seems unneccessarily long. Drop "Poromechanics"?

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 10 above)
Comment (keileg) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:801]: I think I see the question now. While I agree with @jhabriel that such code snippets can be extremely useful, I also share @IvarStefansson's worry about maintenance. There is no way we can automatize detection of errors for this code, and in practice this will soon lead to more than a slight maintenance overhead.

My suggestion is that if this is important, and in this case it may very well be, it should enter a tutorial, which, btw, need not be very long.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 5 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py]: Done.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 10 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py:801]: Ok. I will write the tutorials when time permits.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py (hunk 15 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:112]: Indeed. Fixed.

File: src/porepy/applications/verification_setups/manu_poromech_nofrac.py (hunk 16 above)
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_poromech_nofrac.py]: Done.

File: src/porepy/applications/verification_setups/manu_flow_incomp_frac.py [hunk 17]
```diff
@@ -0,0 +1,807 @@
...
+            approx_matrix_pressure=approx_matrix_pressure,
+            error_frac_flux=error_frac_flux,
+            error_frac_pressure=error_frac_pressure,
+            error_intf_flux=error_intf_flux,
+            error_matrix_flux=error_matrix_flux,
+            error_matrix_pressure=error_matrix_pressure,
+            exact_frac_flux=exact_frac_flux,
+            exact_intf_flux=exact_intf_flux,
+            exact_matrix_flux=exact_matrix_flux,
+            exact_matrix_pressure=exact_matrix_pressure,
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/porepy/applications/verification_setups/manu_flow_incomp_frac.py:205]: Indeed. Fixed.

**Review by IvarStefansson — APPROVED** [review]
